### ISSUE TITLE
Container random port change when added to another network

### ISSUE THREAD

**Issue by ggerbaud** [docker/docker#22836]
I have an annoying issue with random ports (i.e. `docker run -P ...`)

`docker ps` and `docker port` show a port that is not the real opened port.
After some tests I found that the port is changed when I use `docker network connect`

---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------

**Output of `docker version`:**

```
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 813
Server Version: 1.11.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 961
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 3.13.0-37-generic
Operating System: Ubuntu 14.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.58 GiB
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Steps to reproduce the issue:**
```
docker network create net1
docker run --name ctn1 -d --net net1 --net-alias titi -p 8080:8080 tomcat:7
docker network create net01234567
docker run --name ctn2 -d --net net1 -P tomcat:7
docker network connect --alias=toto net01234567 ctn2
```
1. I use tomcat but it could be any image with exposed ports
2. Not sure it helps but if the name of my second network is shorter, it works

**Describe the results you received:**
```
➜  docker port ctn2                      
8080/tcp -> 0.0.0.0:32913
```
```
➜  sudo netstat -ntlp | grep docker-proxy
tcp6       0      0 :::8080                 :::*                    LISTEN      21540/docker-proxy
tcp6       0      0 :::32914                :::*                    LISTEN      21814/docker-proxy
```

**Describe the results you expected:**
The `docker port` output and the real opened port given by `netstat` should have been the same

**Comment by vdemeester** [comment]
Reproduced on a 1.11 host, ping @mavenugo :angel:

**Comment by mavenugo** [comment]
@ggerbaud thanks for reporting the issue. Yes, the gateway resolution logic seems to be re-mapping the port-mapping to the new random port when the gateway network changes.
FYI, if there are multiple networks attached to a container, the gateway resolution logic uses network-name to resolve the preferred gateway. In this case, I think `net0` is prioritized over `net1` and hence the port-mapping changes.

cc @aboch . I thought the move to using Sandbox as the port-mapping resolver should have solved this issue. I guess that is not the case.

**Comment by aboch** [comment]
@mavenugo Yes, when the container's default gateway changes a new mapping is performed. If the host port was not explicitly chosen during docker run (iow, you are fine with a random host port), then after the new mapping the host port will still be random, you could get the same but no guarantee.

The bug as @ggerbaud explained is that docker port o/p is out of sync. I will look into it.

**Comment by thaJeztah** [comment]
@aboch is there a relation with https://github.com/docker/docker/issues/21699 and https://github.com/docker/docker/issues/22325 ?

**Comment by aboch** [comment]
@thaJeztah  I don't think so because those are seen on 1.9.1 and no default gw change was happening there.

**Comment by ggerbaud** [comment]
@aboch right, in my case I'm fine with any random port. Just need to retrieve it somehow :)

**Comment by aboch** [comment]
@ggerbaud I found where the problem is. We will get it fixed in 1.12.
